### PR TITLE
add ability to configure sort_nodeset_fields

### PR DESCRIPTION
**Pull request by mkangia**
https://manage.dimagi.com/default.asp?255102

Good to review 🏛 . 

"Hold off on merge" for any feedback on changes. Deployed on staging.
cc: buddy @biyeun @sheelio

### CODE REVIEW

**Comment by biyeun** [comment]
looks ok but flagging @orangejenny too

File: corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html [hunk 1]
```diff
@@ -116,6 +116,16 @@
 
 {% include 'app_manager/v1/partials/case_list_filtering.html' %}
 
+{{ request|toggle_tag_info:"DETAIL_LIST_TAB_NODESETS" }}
+{% if request|toggle_enabled:'DETAIL_LIST_TAB_NODESETS' %}
+<div data-bind="with: shortScreen">
+    <input data-bind="checked: sortNodesetColumns" type="checkbox">
+    <span>
+        {% trans "Sort entries in the Nodeset tabs" %}
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html]: Minor, but it'd be a ux benefit to make this a label so users can click on the text to check the checkbox. [This](http://getbootstrap.com/css/#forms-controls) has the typical markup for a label+checkbox in bootstrap.

File: corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html [hunk 2]
```diff
@@ -116,6 +116,16 @@
 
 {% include 'app_manager/v1/partials/case_list_filtering.html' %}
 
+{{ request|toggle_tag_info:"DETAIL_LIST_TAB_NODESETS" }}
+{% if request|toggle_enabled:'DETAIL_LIST_TAB_NODESETS' %}
+<div data-bind="with: shortScreen">
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html]: Surprised to see this in the case list tab, I'd imagined adding a "Sorting" section to the case detail and putting it there, since it affects long details not short ones. What would you think of moving it?

File: corehq/apps/app_manager/detail_screen.py [hunk 3]
```diff
@@ -98,6 +100,10 @@ def __init__(self, app, module, detail, column, sort_element=None,
         self.sort_element = sort_element
         self.order = order
         self.id_strings = id_strings
+        self.parent_tab_nodeset = parent_tab_nodeset
+
+    def add_sort_node_for_nodeset_field(self):
```
Comment (orangejenny) [inline corehq/apps/app_manager/detail_screen.py]: Minor, but the name here confused me, `add` sounds like this function alters the column. Maybe `should_sort_nodeset_detail_tab` or `has_sort_node_for_nodeset_detail`?

File: corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html (hunk 1 above)
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html]: This would also be a good place for a help icon, to remind people how this works (that it sorts by the first column, and in what order it sorts).

File: corehq/apps/app_manager/models.py [hunk 4]
```diff
@@ -2030,6 +2030,10 @@ def from_json(cls, data):
                     item.value[lang] = interpolate_media_path(path)
         return to_ret
 
+    @property
```
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: I don't think this is adding much value, since we already expose the string values for detail column formats outside of models.py (app_strings.py and detail-screen-config.js use them).

**Comment by orangejenny** [comment]
@mkangia Could you summarize what the final functionality of this is? [This](https://github.com/dimagi/commcare-hq/pull/16742/files#diff-5fb1ec6df6f30bf5f8223273482f0704R365) looks like it'll be sorted by any search-only fields in the detail, but I thought there was also going to be something about sorting by the first column.

This would also be good info to have in a docstring, probably in `get_nodeset_sort_elements`.

File: corehq/apps/app_manager/util.py [hunk 5]
```diff
@@ -572,3 +572,11 @@ def get_form_data(domain, app):
         module_meta['forms'] = forms
         modules.append(module_meta)
     return modules, errors
+
+
+def sort_nodeset_columns_for_detail(detail_type, detail):
```
Comment (orangejenny) [inline corehq/apps/app_manager/util.py]: Could this be a method on the detail class? It's pretty specific to be in a utils file.

File: corehq/apps/app_manager/suite_xml/sections/details.py [hunk 6]
```diff
@@ -358,6 +362,23 @@ def get_detail_id_safe(self, module, detail_type):
         return detail_id if detail_id in self.active_details else None
 
 
+def get_nodeset_sort_elements(detail):
+    from corehq.apps.app_manager.models import SortElement
+    sort_elements = []
+    tab_spans = detail.get_tab_spans()
+    for tab in detail.get_tabs():
```
Comment (orangejenny) [inline corehq/apps/app_manager/suite_xml/sections/details.py]: This behaves correctly if there are multiple nodeset tabs, right? (I don't know the sorting code well.)

File: corehq/apps/app_manager/models.py (hunk 4 above)
Comment (mkangia) [inline corehq/apps/app_manager/models.py]: it usually prefer having such checks as methods on objects instead of doing the string comparison at various places. So in case the value is to be changed someday, it needs to be done only at one place. For ex here if 'invisible' is changed to 'hidden'.
This made more sense in ruby where method name would actually be 'invisible?' but i think it still holds value here.

**Comment by mkangia** [comment]
> @mkangia Could you summarize what the final functionality of this is? This looks like it'll sort by any search-only fields in the detail, but I thought there was also going to be something about sorting by the first column.

@orangejenny it will sort for any 'Search Only' fields added under nodeset. The check for nodeset is when it is called [here](https://github.com/dimagi/commcare-hq/pull/16742/files/eeb94c9b2e30b822be622397dc29a581f9e739bd#diff-ae61842a4e76ffee74843e118b718a68R446) and not in `get_nodeset_sort_elements` since we don't have nodeset available in that method.

From what i saw in the code base, Sorting by the first column happens in case there are no sort columns by [this method](https://github.com/dimagi/commcare-hq/pull/16742/files/eeb94c9b2e30b822be622397dc29a581f9e739bd#diff-5fb1ec6df6f30bf5f8223273482f0704L361). I don't think we need that here.

File: corehq/apps/app_manager/suite_xml/sections/details.py (hunk 6 above)
Comment (mkangia) [inline corehq/apps/app_manager/suite_xml/sections/details.py]: Yes it does. I did test it. Why do you doubt it?

Comment (orangejenny) [inline corehq/apps/app_manager/suite_xml/sections/details.py]: Was just asking, I like to double-check things :)  Happy you tested it.

Comment (mkangia) [inline corehq/apps/app_manager/suite_xml/sections/details.py]: Oh yes.. no problem.. wanted to just check if something didn't look right. :)

File: corehq/apps/app_manager/models.py (hunk 4 above)
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: Okay, this is basically a stylistic difference of opinion, works for me.

**Comment by orangejenny** [comment]
@mkangia cool, thanks for the explanation

File: corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html (hunk 2 above)
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html]: i moved this under Case Details. The changes in v1 and v2 are slightly different. Had to add a spacer div under v2 and may be we have discontinued using that for spacing now.

File: corehq/apps/app_manager/detail_screen.py (hunk 3 above)
Comment (mkangia) [inline corehq/apps/app_manager/detail_screen.py]: I used "add" because it decides if sort node is to be "added" for the column under nodeset. In column under nodeset we would want to add sort node if needed. So its not for the tab or the detail but for the column under it, hence referring to tab or nodeset detail in that name does not look right.
Can use `add_sort_node_for_nodeset_column` or `has_sort_node_for_nodeset_column` if that makes it better.

**Comment by mkangia** [comment]
@orangejenny All changes done. Good for review.

File: corehq/apps/app_manager/detail_screen.py (hunk 3 above)
Comment (orangejenny) [inline corehq/apps/app_manager/detail_screen.py]: `has_sort_node_for_nodeset_column` sounds good to me

**Review by orangejenny — APPROVED** [review]
Looks good, thanks for polishing up the UI!
